src/coap_openssl.c

```diff
@@ -1366,6 +1514,116 @@ tls_server_name_call_back(SSL *ssl,
   SSL_set_session_secret_cb(ssl, tls_secret_call_back, arg);
   return SSL_TLSEXT_ERR_OK;
 
+error:
+  return SSL_TLSEXT_ERR_ALERT_WARNING;
+}
+
+/*
+ * During the SSL/TLS initial negotiations, psk_tls_server_name_call_back() is
+ * called to see if SNI is being used.
+ *
+ * Set up by SSL_CTX_set_tlsext_servername_callback()
+ * in coap_dtls_context_set_spsk()
+ */
+static int
+psk_tls_server_name_call_back(SSL *ssl,
+                          int *sd UNUSED,
+                          void *arg
+) {
+  coap_dtls_spsk_t *setup_data = (coap_dtls_spsk_t*)arg;
+
+  if (!ssl) {
+    return SSL_TLSEXT_ERR_NOACK;
+  }
+
+  if (setup_data->validate_sni_call_back) {
+    /* SNI checking requested */
+    coap_session_t *c_session = (coap_session_t*)SSL_get_app_data(ssl);
+    coap_openssl_context_t *o_context =
+                  ((coap_openssl_context_t *)c_session->context->dtls_context);
+    const char *sni = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
+    size_t i;
+    char lhint[COAP_DTLS_HINT_LENGTH];
+
+    if (!sni || !sni[0]) {
+      sni = "";
+    }
+    for (i = 0; i < o_context->psk_sni_count; i++) {
+      if (!strcmp(sni, (char*)o_context->psk_sni_entry_list[i].sni.s)) {
+        break;
+      }
+    }
+    if (i == o_context->psk_sni_count) {
+      SSL_CTX *ctx;
+      coap_str_const_t lsni;
+      lsni.length = strlen(sni);
+      lsni.s = (const uint8_t*)sni;
+      const coap_dtls_spsk_info_t *new_entry =
+         setup_data->validate_sni_call_back(&lsni,
+                                            c_session,
+                                            setup_data->sni_call_back_arg);
+      if (!new_entry) {
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      }
+      /* Need to set up a new SSL_CTX to switch to */
+      if (c_session->proto == COAP_PROTO_DTLS) {
+        /* Set up DTLS context */
+        ctx = SSL_CTX_new(DTLS_method());
+        if (!ctx)
+          goto error;
+        SSL_CTX_set_min_proto_version(ctx, DTLS1_2_VERSION);
+        SSL_CTX_set_app_data(ctx, &o_context->dtls);
+        SSL_CTX_set_read_ahead(ctx, 1);
+        SSL_CTX_set_cipher_list(ctx, "TLSv1.2:TLSv1.0");
```

Comment:
IIRC there was a PR that has introduced a pre-processor symbol for the cipher suites to use and have decided to drop TLS 1.0.

Reply:
Good catch - Updated